clusterstate/provider/azure.go

```diff
@@ -0,0 +1,94 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2018-04-01/compute"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/giantswarm/e2e-harness/pkg/framework"
+	azureclient "github.com/giantswarm/e2eclients/azure"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	VirtualMachineSize = "Standard_D4s_v3"
+)
+
+type AzureConfig struct {
+	AzureClient   *azureclient.Client
+	HostFramework *framework.Host
+	Logger        micrologger.Logger
+
+	ClusterID string
+}
+
+type Azure struct {
+	azureClient   *azureclient.Client
+	hostFramework *framework.Host
+	logger        micrologger.Logger
+
+	clusterID string
+}
+
+func NewAzure(config AzureConfig) (*Azure, error) {
+	if config.AzureClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.AzureClient must not be empty", config)
+	}
+	if config.HostFramework == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.HostFramework must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	if config.ClusterID == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.ClusterID must not be empty", config)
+	}
+
+	a := &Azure{
+		azureClient:   config.AzureClient,
+		hostFramework: config.HostFramework,
+		logger:        config.Logger,
+
+		clusterID: config.ClusterID,
+	}
+
+	return a, nil
+}
+
+func (a *Azure) RebootMaster() error {
+	resourceGroupName := a.clusterID
+	scaleSetName := fmt.Sprintf("%s-master", a.clusterID)
+	instanceIDs := &compute.VirtualMachineScaleSetVMInstanceIDs{
+		InstanceIds: to.StringSlicePtr([]string{
+			"0",
```

Comment:
I am not sure we can always reliably assume this is `0`. When somebody changes the overcommitment settings for the VMSS this breaks and gets flappy. It would be safer to just fetch the instance ID and use it accordingly.

Reply:
Ack. I didn't realise this could change. Cool I'll do that.

Reply:
Done.